File: 0201-0300/0250-Count Univalue Subtrees/0250-Count Univalue Subtrees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countUnivalSubtrees(TreeNode* root) {
        int count = 0;
        isUnivalSubtree(root, count);
        return count;
    }
    
private:
    bool isUnivalSubtree(TreeNode* root, int& count) {
        if (root == nullptr) {
            return true;
        }
        
        bool left = isUnivalSubtree(root->left, count);
        bool right = isUnivalSubtree(root->right, count);
        if (left && right &&
           ((root->left == nullptr && root->right == nullptr) ||
            (root->left == nullptr && root->right && root->val == root->right->val ||
            (root->left && root->right == nullptr && root->val == root->left->val) ||
            (root->left && root->right && root->val == root->left->val && root->val == root->right->val)))) {
            ++count;
            return true;
        }
        return false;
    }
};

class Solution {
public:
    int countUnivalSubtrees(TreeNode* root) {
        int count = 0;
        isUnivalSubtree(root, count);
        return count;
    }
    
private:
    bool isUnivalSubtree(TreeNode* root, int& count) {
        if (root == nullptr) {
            return true;
        }
        
        bool left = isUnivalSubtree(root->left, count);
        bool right = isUnivalSubtree(root->right, count);
        if (left && right) {
            if (root->left && root->val != root->left->val) {
                return false;
            }
            
            if (root->right && root->val != root->right->val) {
                return false;
            }
            
            ++count;
            return true;
        }

        return false;
    }
};

File: 0201-0300/0250-Count Univalue Subtrees/0250-Count Univalue Subtrees.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def countUnivalSubtrees(self, root: TreeNode) -> int:
        count = 0
        def isUnivalSubtree(root):
            if root is None:
                return True
            left = isUnivalSubtree(root.left)
            right = isUnivalSubtree(root.right)
            if left and right:
                if root.left and root.val != root.left.val:
                    return False
                if root.right and root.val != root.right.val:
                    return False
                nonlocal count
                count += 1
                return True
            return False
        isUnivalSubtree(root)
        return count
